### ISSUE TITLE
Support for Multi-Tenancy Environments

### ISSUE THREAD

**Issue by s14r** [oauth2-proxy/oauth2-proxy#1863]
# Purpose of this issue
OAuth2 Proxy is currently lacking support for multi-tenancy environments. If you want to secure an application which requires to use different IDP (configurations) based on "something", OAuth2 Proxy is not something which you can use easily.

Even though we have a strong opinion on what we need and how we would approach it, we would love to see additional thoughts and requirements we may able to cover :)

***Disclaimer**: We are not maintaining the project in any way, we just want to start a discussion on how we could include the feature (as we need it in our company) in the most helpful way for the project and the community.*

# Multi-Tenancy in general
There are many different ways on how "tenant identification" happens – we mostly use (sub-)domain based identification while others may use headers based, route based or cookie based identification (or even more?). We think OAuth2 Proxy should have an extensible matching system which can be combined in different ways.

As our use-case is domain driven identification, we would like to start with this in the first place while maintaining an easy-to-extend interface for additional "matchers".

## Thoughts
* Our investigation and plans are based on the [Alpha Config](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/alpha-config) feature, not sure how stable this is but it looks more convenient to us
* CLI options may not be sufficient anymore to cover the additional config complexity
* Config Templates like mentioned in #739 may not be enough, specifically when it comes to different providers for different tenants
* Config structure should be open for pre-planned modifications to not force users to change their config from release to release (e.g. when new matchers are introduced, etc.)

## Config
Here is a very *rough* idea of how a configuration could look like in the future.
We would make use of the already existing [Alpha Config](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/alpha-config) while we would spent time to harmonize configuration.
If we go this way, it would definitely lead to a breaking change release as configuration API would change entirely.

There is already a helper to create an alpha config out of the existing one and we think it would make sense to keep this approach to have a "one-shot" configuration migration.

```yaml

server:
  bindAddress: 0.0.0.0:4180
  emailDomains: "*"
  cookieSecure: true
  skipProviderButton: false
  redirectURL: "/oauth2/callback"
  skipAuthPreflight: true
  cookieRefresh: 15m
  defaultProvider: keycloak-tenant-1
  
  # ... more options ...
  
upstreamServers:
  - id: api
    flushInterval: 1s
    passHostHeader: true
    proxyWebSockets: true
    timeout: 30s
    uri: http://127.0.0.1
    path: /

providers:
  keycloak-tenant-1:
    provider: keycloak
    keycloakConfig:
      clientID: my-keycloak-client-id-1
      clientSecret: my-keycloak-client-secret-1
      loginURL: "https://<keycloak host>/auth/realms/<your realm>/protocol/openid-connect/auth"
      redeemURL: "http(s)://<keycloak host>/auth/realms/<your realm>/protocol/openid-connect/token"
      profileURL: "http(s)://<keycloak host>/auth/realms/<your realm>/protocol/openid-connect/userinfo"
      validateURL: "http(s)://<keycloak host>/auth/realms/<your realm>/protocol/openid-connect/userinfo"
      keycloakGroups: 
        - group1
        - group2
  
  keycloak-tenant-2:
    provider: keycloak
    keycloakConfig:
      clientID: my-keycloak-client-id-2
      clientSecret: my-keycloak-client-secret-2
      loginURL: "https://<keycloak host>/auth/realms/<your realm>/protocol/openid-connect/auth"
      redeemURL: "http(s)://<keycloak host>/auth/realms/<your realm>/protocol/openid-connect/token"
      profileURL: "http(s)://<keycloak host>/auth/realms/<your realm>/protocol/openid-connect/userinfo"
      validateURL: "http(s)://<keycloak host>/auth/realms/<your realm>/protocol/openid-connect/userinfo"
      keycloakGroups:
        - group1
        - group2

# Matching defines in which way the OAuth2 Proxy will determine which provider to use on each request
matcher:
  - provider: keycloak-tenant-1
    matchers:
      # if ONE of the matchers hit the given provider is used for the auth request  
      - type: domain
        domain: "tenant1.my-great.app"

      # so you can combine additional matching methods to support different scenarios 
      - type: header
        header: "X-TENANT-ID"
        value: "tenant-1"
        
    # grouping could potentially be used to support more specific use-cases
    - provider: keycloak-tenant-2
        type: group
        group:
          # both conditions need to fulfilled in order to match
          # this is basically an "AND" operation
          - type: domain
            domain: "tenant2.my-great.app"
          - type: header
            header: "X-TENANT-ID"
            value: "tenant-2"
```

I'm happy to have @beezerk23 here with me who already started working on this feature internally. 
Happy to hear your thoughts!

Hey @JoelSpeed, please jump in if we've overseen something :)

Looking forward and have a good one guys!

**Comment by salmanazmat666** [comment]
Hello @s14r,
I have some comments and questions related to architecture.

1. How would we add and remove tenants dynamically? In our use case, we have one keycloak realm per tenant and we'd be adding tenants as new customers come on-board and this would be a very frequent occurrence. With this architecture, we'll have to update the configuration file and then restart the containers/pods, which I think is not very efficient.
2. Having one matcher per tenant, I guess you'd be doing something like following in every request's handler.
```
for i := 0; i < len(matchers); i++ {
  if matchers[i].Match(req) {
    matchers[i].Provider.Handle()
  }
}
```
If we have tenants in some 1000s, we'd be trying to match 1000s of matchers in each request, which again I think, wouldn't scale efficiently.

Looking forward to hear from you. Do these issues affect your use case ?

**Comment by salmanazmat666** [comment]
Hello @s14r,
Thanks for the reply.

About the providers, I'm still not very comfortable with the config hot reload. I'm more inclined towards storing the providers in a database, implementing a cache over it for fast read/write access and then exposing a simple API for CRUD operations on providers.
But at the same time, I don't want to request too much from your side that'll distract you from your goals in your timeframe. So, I'd just request if you can hide your implementation of loading providers from configuration file, hot reload and related things behind a simple facade interface like following, it will really help everyone.

```
type ProviderGetter interface{
   GetProvider(id string) *Provider
}
```

Same goes for matchers, it'd be best if you can use interface there as well, so we can improve the implementation in future or even add a new implementation without affecting the core oauth2 logic. We can use something similar to the following interfaces and then use that everywhere in code,

```
type Matcher interface{
  MatchProvider(*http.Request, *sessionsapi.SessionState) string //returns provider id
}

or

type Matcher interface{
  MatchProvider(*http.Request, *sessionsapi.SessionState) *Provider //returns provider
}
```

If that seems reasonable, let's have a discussion around locking down the interface definitions.

Thanks,

**Comment by salmanazmat666** [comment]
Hi @s14r,

Waiting for your comments on the above. Also, please let me know if there's a possibility that I can help with the development. I'll be happy to help.

Regards,
Salman

**Comment by beezerk23** [comment]
Hey @salmanazmat666 ,

as @s14r mentioned im working on this for our company. 

Having a database connection is in our opinion too much overhead. Imagine having the proxy in front of a simple frontend application where you not already have a database. We want to keep the simplicity of the oauth2-proxy by not making it depending on something like a database. However, we also want to make the matcher easily extendable, so an interface definitely makes sense here. Also i'll think of having an interface for getting the provider, but currently we need to focus a bit on our needs so this is probably something for the future.

Im also member of the Gophers Slack Workspace, if you want you can reach me there :) 

Hope this answers your question.

**Comment by beezerk23** [comment]
Hey @laurivosandi ,

what exactly should we not implement? We try to keep it as simple as possible all the time, so you can(!) configure multiple providers and use the suggested matching configs OR you still can just configure one provider which is used directly. 

As mentioned above we will not implement any database connection as i agree with you that its an unnecessary complexity layer. 

Adding and removing IdPs dynamically will also not be a part of the oauth proxy. We want to stick with the configuration file and loading it once.

**Comment by salmanazmat666** [comment]
@beezerk23 @s14r 
Since we were also implementing this feature internally, I have raised a PR that meets your requirements of loading tenants from the config file and has provision to be extended easily to add/remove tenants dynamically without touching the main Oauth2 code. 

I have submitted a PR #1923, you are welcome to have a look at it and comment what you think. We are also open to collaboration with you guys on this feature.
